deploy_heroku.py

```diff
@@ -3,61 +3,128 @@
+        headers=headers,
+        data=json.dumps(payload).encode() if payload is not None else None,
+    )
+
+
+def get_latest_commit_from_slug(app: str, api_key: str, slug_id: str):
+    request = get_request(app, api_key, Endpoint.SLUGS.format(slug_id))
+
+    with urllib.request.urlopen(request) as response:
+        assert response.getcode() == 200, response.getcode()
+        return json.loads(response.read())["commit"]
 
 
 def get_latest_heroku_release(app: str, api_key: str) -> HerokuRelease:
-    request = urllib.request.Request(f"https://api.heroku.com/apps/{app}/releases")
-    for key, value in (
-            ("Accept", "application/vnd.heroku+json; version=3"),
-            ("Authorization", f"Bearer {api_key}"),
-            ("Content-Type", "application/json"),
-            ("Range", "version ..; order=desc,max=10;"),
-    ):
-        request.add_header(key, value)
+    request = get_request(app, api_key, Endpoint.RELEASES)
 
     with urllib.request.urlopen(request) as response:
         assert response.getcode() in (200, 206), response.getcode()
         latest_release, *_ = json.loads(response.read())
-        return HerokuRelease(**{prop.name: latest_release[prop.name] for prop in dataclasses.fields(HerokuRelease)})
+        return release_from_response(latest_release)
 
 
-def deploy_heroku_command(commit_hash: str, api_key: str, app: str) -> List[str]:
-    return ["git", "push", f"https://heroku:{api_key}@git.heroku.com/{app}.git", f"{commit_hash}:refs/heads/master"]
+def trigger_release_retry(app: str, api_key: str, release: HerokuRelease):
+    request = get_request(
+        app=app,
+        api_key=api_key,
+        endpoint=Endpoint.RELEASES,
+        payload={"slug": release.slug_id, "description": f"Retry of v{release.version}: {release.description}"},
+    )
 
+    with urllib.request.urlopen(request) as response:
+        assert response.getcode() == 201, response.getcode()
+        return release_from_response(json.loads(response.read()))
 
-if __name__ == "__main__":
-    parser = argparse.ArgumentParser(description="perform deployment to Heroku")
-    parser.add_argument("-a", "--app", default=os.environ.get("APP"), type=str)
-    parser.add_argument("-K", "--api-key", default=os.environ.get("API_KEY"), type=str)
-    parser.add_argument("-c", "--commit-hash", default=os.environ.get("COMMIT_HASH"), type=str)
-    parser.add_argument("--dry-run", action="store_true")
-    args = parser.parse_args()
 
-    heroku_app, heroku_api_key, git_commit_hash = args.app, args.api_key, args.commit_hash
-    assert heroku_app and heroku_api_key and git_commit_hash, "app, API key and commit hash required!"
+def deploy_heroku_command(commit_hash: str, api_key: str, app: str, rollback: bool) -> str:
+    git_url = f"https://heroku:{api_key}@git.heroku.com/{app}.git"
+    return f"git push {git_url} {commit_hash}:refs/heads/master {'--force' if rollback else ''}".strip()
```

Comment:
```suggestion
    return f"git push {git_url} {commit_hash}:refs/heads/master {'--force' if rollback else ''}"
```

Eigentlich `strip` hier ist gar nicht gebraucht oder?

Reply:
Habe ich hinzugeügt, da sonst ein Leerzeichen am Ende ist, was sonst bei den Tests doof aussah.

Reply:
Alles klar, ich kann damit leben - habe mir selbst dann später gedacht, dass du es nicht "unsauber" in Tests haben wolltest...